pkg/controller/drain.go

```diff
@@ -373,6 +387,236 @@ func (o *DrainOptions) evictPods(pods []api.Pod, policyGroupVersion string, getP
+	returnChInt := make(chan error, 1)
+	detachCh := make(chan bool, 1)
+
+	sortPodsByPriority(pods)
+
+	volMap := o.doAccountingOfPvs(pods)
+	filterSharedPVs(volMap)
+
+	for {
+		retryPods := []corev1.Pod{}
+		for _, pod := range pods {
+			pvs := volMap[pod.Namespace+"/"+pod.Name]
+			go o.evictPodInternal(pod, policyGroupVersion, getPodFn, returnChInt, true)
+			startTime := time.Now()
+			select {
+			case err := <-returnChInt:
+				if apierrors.IsTooManyRequests(err) {
+					glog.V(3).Info("Pod ", pod.Namespace, "/", pod.Name, " couldn't be evicted because of PDB violation. Will be retried.")
+					// Pod eviction failed because of PDB violation, retry
+					retryPods = append(retryPods, pod)
+					continue
+				} else if err != nil {
+					returnCh <- fmt.Errorf("Error when evicting pod: %v/%v. Err: %v", pod.Namespace, pod.Name, err)
+					continue
+				}
+
+				go o.waitForDetach(pvs, o.nodeName, detachCh)
+
+				glog.V(3).Info("waiting pv to detach for pod: ", pod.Name)
+				select {
+				case <-detachCh:
+					glog.V(3).Info("Detached pv for pod: ", pod.Name)
+					returnCh <- nil
+				case <-time.After(o.PvDetachTimeout - time.Now().Sub(startTime)):
+					glog.Warningf("Timout occured when detaching pv for pod: %v", pod.Name)
+					returnCh <- nil
+				}
+			case <-time.After(o.PvDetachTimeout):
+				returnCh <- fmt.Errorf("Timout when evicting pod: %v", pod)
+			}
+		}
+		if len(retryPods) == 0 {
+			break
+		}
+		glog.V(4).Info("Eviction for some pods will be retried after 5 seconds.")
+		pods = retryPods
+		time.Sleep(5 * time.Second)
+	}
+	return
+}
+
+func (o *DrainOptions) getPvs(pod corev1.Pod) []string {
+	pvs := []string{}
+	for _, vol := range pod.Spec.Volumes {
+		if vol.PersistentVolumeClaim != nil {
+			pvc, err := o.client.CoreV1().PersistentVolumeClaims(pod.Namespace).Get(vol.PersistentVolumeClaim.ClaimName, metav1.GetOptions{})
+			if err != nil {
+				// If error is not nil, then skip this volume
+				continue
+			}
+			pvs = append(pvs, pvc.Spec.VolumeName)
+		}
+	}
+	return pvs
+}
+
+func (o *DrainOptions) waitForDetach(pvNames []string, nodeName string, detachCh chan bool) {
+	if pvNames == nil || len(pvNames) == 0 || nodeName == "" {
+		// If pv or node name is not available, nothing to do. Just log this as warning
+		glog.Warningf("Node name: %v, list of PVs: %v", nodeName, pvNames)
+		detachCh <- false
+		return
+	}
+	volumeIDs, err := o.getVolIDsFromDriver(pvNames)
+	if err != nil {
+		glog.Errorf("Error when getting volume IDs from the driver. Err: %v", err)
+		detachCh <- false
+		return
+	}
+	glog.V(4).Info("Volume names from driver: ", volumeIDs)
+
+	found := true
+	for found {
+		found = false
+		node, err := o.client.Core().Nodes().Get(nodeName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			glog.V(4).Info("Node not found: ", nodeName)
+			detachCh <- true
```

Comment:
This is technically not true. It might be better to make `detachCh` `chan error` and emit error here and log it on the consuming side.